python/dgl/nn/pytorch/explain/gnnexplainer.py

```diff
@@ -425,3 +398,395 @@ def explain_graph(self, graph, feat, **kwargs):
+    model : nn.Module
+        The GNN model to explain.
+        * The required arguments of its forward function are graph and feat.
+          The latter one is for input node features.
+        * It should also optionally take an eweight argument for edge weights
+          and multiply the messages by it in message passing.
+        * The output of its forward function is the logits for the predicted
+          node/graph classes.
+        See also the example in :func:`explain_node` and :func:`explain_graph`.
+    num_hops : int
+        The number of hops for GNN information aggregation.
+    lr : float, optional
+        The learning rate to use, default to 0.01.
+    num_epochs : int, optional
+        The number of epochs to train.
+    alpha1 : float, optional
+        A higher value will make the explanation edge masks more sparse by decreasing
+        the sum of the edge mask.
+    alpha2 : float, optional
+        A higher value will make the explanation edge masks more sparse by decreasing
+        the entropy of the edge mask.
+    beta1 : float, optional
+        A higher value will make the explanation node feature masks more sparse by
+        decreasing the mean of the node feature mask.
+    beta2 : float, optional
+        A higher value will make the explanation node feature masks more sparse by
+        decreasing the entropy of the node feature mask.
+    log : bool, optional
+        If True, it will log the computation process, default to True.
+    """
+    
+    def __init__(self, 
+                 model, 
+                 num_hops, 
+                 lr=0.01, 
+                 num_epochs=100, 
+                 *,
+                 alpha1=0.005,
+                 alpha2=1.0,
+                 beta1=1.0,
+                 beta2=0.1,
+                 log=True):
+        super(HeteroGNNExplainer, self).__init__()
+        self.model = model
+        self.num_hops = num_hops
+        self.lr = lr
+        self.num_epochs = num_epochs
+        self.alpha1 = alpha1
+        self.alpha2 = alpha2
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.log = log
+
+    def _init_masks(self, graph, feat):
+        r"""Initialize learnable feature and edge mask.
+        Parameters
+        ----------
+        graph : DGLHeteroGraph
+            Input graph.
+        feat : Dictionary of Tensor
+            Input node features.
+        Returns
+        -------
+        feat_masks : Dictionary of Tensor
+            Dictionary of { ntypes: features }
+            Feature mask of shape :math:`(1, D)`, where :math:`D`
+            is the feature size.
+        edge_masks : Dictionary of Tensor
+            Dictionary of { canonical_etypes: features }
+            Edge mask of shape :math:`(E)`, where :math:`E` is the
+            number of edges.
+        """
+        device = graph.device
+        feat_masks = {}
+        for node_type, feature in feat.items():
+            if len(feature.size()) == 1:
```

Comment:
The feature.size() is just to make the model more flexible in handling the input so that you can pass a feature of length [node_size] instead of having to pass it as [node_size, 1]. I can remove the check and initialize it as before, as removing this check has no effect.

Reply:
I think typically `feature` is a 2D tensor/matrix. We do not need to handle cases of 1D tensors for simplicity unless there will be many user requests.

Reply:
okay, understood.